Sources/JWTKit/JWTSigners.swift

```diff
@@ -108,8 +109,99 @@ public final class JWTSigners {
+        _ token: Message,
+        as payload: Payload.Type = Payload.self,
+        rootCert: Message
+    ) throws -> Payload
+        where Message: DataProtocol, Payload: JWTPayload
+    {
+        let parser = try JWTParser(token: token)
+        let header = try parser.header()
+        guard let x5c = header.x5c else {
+            throw JWTError.generic(identifier: "JWS", reason: "No x5c certificates provided")
+        }
+        
+        // Verify the chain
+        // The first cert is used to sign the JWS
+        // Each subsequent cert should be used to certify the previous one
+        // For the last cert we use the root cert passed in
+        // https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.6
+        for (index, certificate) in x5c.enumerated() {
+            if index == 0 {
+                continue
+            }
+            
+            if index == x5c.count {
+                let rootCertx509 = try ECDSAKey.load(pem: rootCert) { bio in
+                    CJWTKitBoringSSL_PEM_read_bio_X509(bio, nil, nil, nil)
+                }
+                defer { CJWTKitBoringSSL_X509_free(rootCertx509) }
+                
+                let certToVerify = addBoundaryToCert(certificate)
+                let certToVerifyX509 = try ECDSAKey.load(pem: [UInt8](certToVerify.utf8)) { bio in
+                    CJWTKitBoringSSL_PEM_read_bio_X509(bio, nil, nil, nil)
+                }
+                defer { CJWTKitBoringSSL_X509_free(certToVerifyX509) }
+                
+                let rootPKey = CJWTKitBoringSSL_X509_get_pubkey(rootCertx509)
+                defer { CJWTKitBoringSSL_EVP_PKEY_free(rootPKey) }
+                
+                guard CJWTKitBoringSSL_X509_verify(certToVerifyX509, rootPKey) == 1 else {
+                    throw JWTError.generic(identifier: "JWS", reason: "Certificate verification failed")
+                }
+            } else {
+                let certString = addBoundaryToCert(certificate)
+                
+                let x509 = try ECDSAKey.load(pem: [UInt8](certString.utf8)) { bio in
+                    CJWTKitBoringSSL_PEM_read_bio_X509(bio, nil, nil, nil)
+                }
+                defer { CJWTKitBoringSSL_X509_free(x509) }
+                
+                let cert2String = addBoundaryToCert(x5c[index-1])
+                let x5092 = try ECDSAKey.load(pem: [UInt8](cert2String.utf8)) { bio in
+                    CJWTKitBoringSSL_PEM_read_bio_X509(bio, nil, nil, nil)
+                }
+                defer { CJWTKitBoringSSL_X509_free(x5092) }
+                
+                let pkey = CJWTKitBoringSSL_X509_get_pubkey(x509)
+                defer { CJWTKitBoringSSL_EVP_PKEY_free(pkey) }
+                
+                guard CJWTKitBoringSSL_X509_verify(x5092, pkey) == 1 else {
+                    throw JWTError.generic(identifier: "JWS", reason: "Certificate verification failed")
+                }
+            }
+        }
```

Comment:
To do cert verification it's better not to reinvent `X509_STORE_CTX`. Prefer using `X509_verify_cert`, which will force you to construct the `X509_STORE_CTX` appropriately for validation.